operator/v1/types_baremetal.go

```diff
@@ -0,0 +1,68 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// Metal3ProvisioningController contains configuration used by the Provisioning service (Ironic)
+// to provision baremetal hosts.
+//
+type Metal3ProvisioningController struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the
+	// Metal3ProvisioningController.
+	Spec Metal3ProvisioningControllerSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the
+	// Metal3ProvisioningController.
+	Status Metal3ProvisioningControllerStatus `json:"status,omitempty"`
+}
+
+// ProvisioningDHCP
+type ProvisioningDHCP struct {
+	// managementState indicates whether the DHCP service would be started within
+	// the metal3 cluster or would use an external DHCP service
+	ManagementState ManagementState `json:"managementState"`
```

Comment:
I think instead of adding the ManagementState fields, we could embed the OperatorSpec struct instead

```suggestion
	OperatorSpec `json:",inline"`
```

See: https://golang.org/doc/effective_go.html#embedding

Reply:
if you need an operatorspec, that sounds good, but please place it directly in spec.

Reply:
I am trying to follow what was agreed upon in https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/baremetal-provisioning-config.md.

I'll add OperatorSpec instead of just adding ManagementState since we could use the additional fields in OperatorSpec.

I am not clear about adding OperatorSpec directly into the Spec struct for the new CR. According to the proposal that just merged, we were going to add another struct just for the DHCP fields : ManagementState (which is now OperatorSpec) and DHCP Range. If we move the OperatorSpec directly to the Spec, then we do not need a separate ProvisioningDHCP struct and add the DHCP range also directly to the Spec.

Reply:
> if you need an operatorspec, that sounds good, but please place it directly in spec.

What do you mean by "directly in the spec"? Why not embed it in ProvisioningDHCP? This is an OperatorSpec for the DHCP service and originates from the discussion we had in the enhancements PR, as it fits that model.

This was also suggested by @soltysh https://github.com/openshift/api/pull/540#discussion_r358303165

Reply:
Yes. Not questioning any of that. There are 2 ways to do this -
Option 1. type ProvisioningDHCP struct {
        OperatorStatus `json:",inline"`
        DHCPRange string `json:"DHCPRange,omitempty"`
}
We had discussed adding a new ProvisioningDHCP struct in the enahncement and now OperatorStatus will be part of it instead of ManagementState.

Option 2: type Metal3ProvisioningSpec struct {
        // provisioningInterface is the name of the network interface on a Baremetal server connected
        // to the provisioning network.
        ProvisioningInterface string `json:"provisioningInterface"`

        // provisioningIP is the IP address assigned to the provisioningInterface for provisioning
        // the baremetal node. This IP address should be within the provisioning subnet, and
        // outside of the DHCP range.
        ProvisioningIP string `json:"provisioningIP"`

        // provisioningNetworkCIDR is the network on which the baremetal nodes are provisioned.
        // The provisioningIP and the IPs in the dhcpRange all come from within this network.
        ProvisioningNetworkCIDR string `json:"provisioningNetworkCIDR"`
        // directly adding OperatorStatus to Metal3ProvisioningSpec would look like this
        OperatorStatus `json:",inline"`
        DHCPRange string `json:"DHCPRange,omitempty"`
}

Is there one of these 2 Options that is preferred over the other?

Reply:
@deads2k I can see why OperatorSpec could be added directly in the spec. In our case, we are using the ManagementState within the OperatorSpec to mean something specific about the state of the DHCP service.  This value combined with the DHCPRange represents the complete DHCP configuration. So, for that reason we really cannot embed it in the Spec and have to leave it within the ProvisioningDHCP struct.